src/runtime-rs/crates/hypervisor/src/device/driver/vhost_user_blk.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2023 Alibaba Cloud
+// Copyright (c) 2023 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+
+use super::VhostUserConfig;
+use crate::{
+    device::{Device, DeviceType},
+    Hypervisor as hypervisor,
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct VhostUserBlkDevice {
+    pub device_id: String,
+    /// If set to true, the drive is opened in read-only mode. Otherwise, the
+    /// drive is opened as read-write.
+    pub is_readonly: bool,
+
+    /// Don't close `path_on_host` file when dropping the device.
+    pub no_drop: bool,
+
+    /// driver type for block device
+    pub driver_option: String,
+
+    pub attach_count: u64,
+    pub config: VhostUserConfig,
+}
+
+impl VhostUserBlkDevice {
+    // new creates a new VhostUserBlkDevice
+    pub fn new(device_id: String, config: VhostUserConfig) -> Self {
+        VhostUserBlkDevice {
+            device_id,
+            attach_count: 0,
+            config,
+            ..Default::default()
+        }
+    }
+}
+
+#[async_trait]
+impl Device for VhostUserBlkDevice {
+    async fn attach(&mut self, h: &dyn hypervisor) -> Result<()> {
+        // increase attach count, skip attach the device if the device is already attached
+        if self
+            .increase_attach_count()
+            .await
+            .context("failed to increase attach count")?
+        {
+            return Ok(());
+        }
+
+        if let Err(e) = h.add_device(DeviceType::VhostUserBlk(self.clone())).await {
+            self.decrease_attach_count().await?;
+
+            return Err(e);
+        }
+
+        return Ok(());
+    }
+
+    async fn detach(&mut self, h: &dyn hypervisor) -> Result<Option<u64>> {
+        // get the count of device detached, skip detach once it reaches the 0
+        if self
+            .decrease_attach_count()
+            .await
+            .context("failed to decrease attach count")?
+        {
+            return Ok(None);
+        }
+
+        if let Err(e) = h
+            .remove_device(DeviceType::VhostUserBlk(self.clone()))
+            .await
+        {
+            self.increase_attach_count().await?;
+
+            return Err(e);
+        }
+
+        Ok(Some(self.config.index))
+    }
+
+    async fn get_device_info(&self) -> DeviceType {
+        DeviceType::VhostUserBlk(self.clone())
+    }
+
+    async fn increase_attach_count(&mut self) -> Result<bool> {
```

Comment:
Can `increase_attach_count` and `decrease_attach_count` be abstracted into public utils, other devices are also needed.

Reply:
Thx, I can't agree more with you. But in this patch, it will be kept as it. Another new PR will do refactor it when other drivers are all ready.